### PR TITLE
Add no-dict-direct-access checker

### PR DESCRIPTION
**Pull request by pmiddend**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

I noticed that I rarely use direct dictionary access, as in `myfunction(foo["bar"])`, since `foo` might not contain `"bar"`. I usually go for `foo.get("bar")` (or some other default value) to be safe. I wanted a checker that checks my code for these direct dictionary subscripts and reports them. I found a few bugs that way already and thought this might fit into pylint in general.

The following will produce a message:

```python
mydict = { "foo": 3 }
print(mydict["foo"])
```

Whereas this will not:

```python
mydict = { "foo": 3 }
print(mydict.get("foo"))
```

This is also fine:

```python
mydict = {}
mydict["foo"] = 3
```

Since assignment is safe per se.

### CODE REVIEW

File: doc/whatsnew/2/2.14/full.rst [hunk 1]
```diff
@@ -5,6 +5,8 @@ What's New in Pylint 2.14.2?
 ----------------------------
 Release date: TBA
 
+* Added new checker ``no-dict-direct-access``.
+
```
Comment (jacobtylerwalls) [inline doc/whatsnew/2/2.14/full.rst:9]: New features go in the next minor version, so this should move to 2.15.rst

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for your interest in contributing this to pylint. I think at the least this should be an optional extension. See the examples in `pylint/checkers/extensions`. So that will solve the bulk of the test failures.

You might wait to hear positive feedback from others before continuing, as this is the sort of thing that might be too strict even for an optional extension.

**Comment by Pierre-Sassoulas** [comment]
I don't have time for a full review right now but as Jacob said it needs to be an extension. We have really opinionated extensions already so it's ok. (While used, for example).

**Comment by pmiddend** [comment]
Thanks @Pierre-Sassoulas for the vote of confidence and @jacobtylerwalls for the general hints! I'll move the check to an extension and am absolutely open to more opinions on the matter. Worst case, I'll release it on pypi or something and it'll be more unofficial. :)

**Comment by pmiddend** [comment]
I have fixed a few issues and put the checker into an extension.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I think this need some functional tests, in particular what happen when the code is:
```python
d = {}
try:
    d["a"] = 1
except KeyError:
    print("The 'a' key does not exists in d")
```

**Comment by jacobtylerwalls** [comment]
Thanks for taking the time to prepare a pull request. I agree, though, with Pierre that this is probably better as a plugin maintained outside the pylint project. If we hear more requests, I'll point folks here and revisit the issue.

**Comment by pmiddend** [comment]
@jacobtylerwalls I understand, and have begun migrating this to a pylint plugin which will be released on pypi. Thanks for the pleasent review process :)

**Pull request closed without merging**